File: 2020/fsharp/AOC/Day15.fs
module AOC.Day15 

let takeTurn (recitation, previous) turn = 
    if Map.containsKey previous recitation then 
        let lastSaid = Map.find previous recitation 
        Map.add previous (turn - 1) recitation, turn - lastSaid - 1 
    else
        Map.add previous (turn - 1) recitation, 0

let runGame data ending = 
    let recitation = Seq.zip data (seq {1 .. (List.length data) - 1}) |> Map.ofSeq
    seq {(List.length data) + 1 .. ending}
    |> Seq.fold takeTurn (recitation, List.last data)
    |> snd

let part1 data = 
    runGame data 2020

let part2 data = 
    runGame data 30000000

File: 2020/julia/AOC/src/day15.jl
module Day15 

function whats_next!(recitation, previous, turn)
	if haskey(recitation, previous)
		last_said = recitation[previous]
		recitation[previous] = turn-1
		turn - last_said - 1
	else
		recitation[previous] = turn-1
		0 
	end
end

function run_game(data, ending)
	recitation = Dict(zip(data[1:end-1], 1:(length(data)-1)))
	previous = data[end]
	for turn in (length(data)+1):ending
		previous = whats_next!(recitation, previous, turn)
	end
	previous
end

part1(;data=[1,0,15,2,10,13]) = run_game(data, 2020)

part2(;data=[1,0,15,2,10,13]) = run_game(data, 30000000)

end